core/shared/src/main/scala/zio/Cause.scala

```diff
@@ -325,24 +309,50 @@ sealed trait Cause[+E] extends Product with Serializable { self =>
       case Meta(c, data)    => c.stripFailures.map(Meta(_, data))
     }
 
-  final def succeeded: Boolean = !failed
-
   final def traces: List[ZTrace] =
     self
       .foldLeft(List.empty[ZTrace]) {
         case (z, Traced(_, trace)) => trace :: z
       }
       .reverse
 
-  private def foldLeft[Z](z: Z)(f: PartialFunction[(Z, Cause[E]), Z]): Z =
-    (f.lift(z -> self).getOrElse(z), self) match {
-      case (z, Then(left, right)) => right.foldLeft(left.foldLeft(z)(f))(f)
-      case (z, Both(left, right)) => right.foldLeft(left.foldLeft(z)(f))(f)
-      case (z, Traced(cause, _))  => cause.foldLeft(z)(f)
-      case (z, Meta(cause, _))    => cause.foldLeft(z)(f)
+  private def find[Z](f: PartialFunction[Cause[E], Z]): Option[Z] = {
+    @tailrec
+    def loop(cause: Cause[E], stack: List[Cause[E]]): Option[Z] =
+      f.lift(cause) match {
```

Comment:
I suggest `f.lift(cause) orElse ...` to avoid pattern matching

Reply:
The issue is if you do that the recursive call to `loop` is no longer in tail position.

Reply:
Yes indeed 👍 I should have put the comment below with this one 🤦‍♂